sdks/java/io/kinesis/src/test/java/org/apache/beam/sdk/io/kinesis/KinesisMockWriteTest.java

```diff
@@ -0,0 +1,278 @@
+    properties.setProperty("KinesisPort", "4567");
+    properties.setProperty("VerifyCertificate", "false");
+
+    Iterable<byte[]> data = Arrays.asList("1".getBytes(), "2".getBytes(), "3".getBytes());
+    p.apply(Create.of(data))
+        .apply(
+            KinesisIO.write()
+                .withStreamName(STREAM)
+                .withPartitionKey(PARTITION_KEY)
+                .withAWSClientsProvider(new FakeKinesisProvider(getMockedAmazonKinesisClient()))
+                .withProducerProperties(properties));
+    p.run().waitUntilFinish();
+
+    assertEquals(3, kinesisService.getAddedRecords().get());
+  }
+
+  @Test
+  public void testWriteAndReadFromMockKinesis() {
+    KinesisServiceMock kinesisService = KinesisServiceMock.getInstance();
+
+    Iterable<byte[]> data = Arrays.asList("1".getBytes(), "2".getBytes());
+    p.apply(Create.of(data))
+        .apply(
+            KinesisIO.write()
+                .withStreamName(STREAM)
+                .withPartitionKey(PARTITION_KEY)
+                .withAWSClientsProvider(new FakeKinesisProvider(getMockedAmazonKinesisClient())));
+    p.run().waitUntilFinish();
+    assertEquals(2, kinesisService.getAddedRecords().get());
+
+    List<List<AmazonKinesisMock.TestData>> testData = kinesisService.getShardedData();
+
+    int noOfShards = 1;
+    int noOfEventsPerShard = 2;
+    PCollection<AmazonKinesisMock.TestData> result =
+        p2.apply(
+                KinesisIO.read()
+                    .withStreamName(STREAM)
+                    .withInitialPositionInStream(InitialPositionInStream.TRIM_HORIZON)
+                    .withAWSClientsProvider(new AmazonKinesisMock.Provider(testData, 10))
+                    .withMaxNumRecords(noOfShards * noOfEventsPerShard))
+            .apply(ParDo.of(new KinesisRecordToTestData()));
+    PAssert.that(result).containsInAnyOrder(Iterables.concat(testData));
+    p2.run().waitUntilFinish();
+  }
+
+  private static final class BasicKinesisPartitioner implements KinesisPartitioner {
+    @Override
+    public String getPartitionKey(byte[] value) {
+      return String.valueOf(value.length);
+    }
+
+    @Override
+    public String getExplicitHashKey(byte[] value) {
+      return null;
+    }
+  }
+
+  private static final class FakeKinesisProvider implements AWSClientsProvider {
+    private transient AmazonKinesis client;
+
+    public FakeKinesisProvider() {
+    }
+
+    public FakeKinesisProvider(AmazonKinesis client) {
+      this.client = client;
+    }
+
+    @Override
+    public AmazonKinesis getKinesisClient() {
+      return client;
+    }
+
+    @Override
+    public AmazonCloudWatch getCloudWatchClient() {
+      throw new RuntimeException("Not implemented");
+    }
+
+    @Override
+    public IKinesisProducer getKinesisProducer(KinesisProducerConfiguration config) {
+      return new KinesisProducerMock(config);
+    }
+  }
+
+  private static class KinesisRecordToTestData
```

Comment:
deduplicate with read part of the io

Reply:
this KinesisRecordToTestData class is both in kinesis write test and in kinessis read test